torch/_inductor/codecache.py

```diff
@@ -575,79 +588,193 @@ def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+    - A given graph could have multiple compiled versions, corresponding to
+      different sets of guards. Therefore, we store cache entries in the form:
+          <temp dir>/<fx graph hash>/<serialized metatdata>
+    - On lookup, we compute the key from the graph details, iterate over all
+      leaf files in the corresponding subdirectory, deserialize the entry, and
+      evaluate its guards expression. If the evaluation succeeds, we have a
+      cache hit. If it fails, we compile the graph and store a new entry.
+    - Finally, on a cache hit, we need to make sure any guards that would
+      have been created during compilation are added to the current context.
     """
 
     # TODO(masnesral): Investigate whether it's beneficial to store compiled graphs
     # in an in-memory cache after loading from disk.
-    @classmethod
-    def save_graph(cls, key: str, compiled_graph: CompiledFxGraph):
+    @staticmethod
+    def _get_subdir(key: str) -> str:
+        """
+        Return the disk location for a given cache key.
+        """
+        return os.path.join(cache_dir(), key[1:3], key)
+
+    @staticmethod
+    def _lookup_graph(
+        key: str,
+        example_inputs: List[torch.Tensor],
+    ) -> Optional[CompiledFxGraph]:
+        """
+        Lookup a compiled graph in the cache by key. On a hit, return the
+        deserialized CompiledFxGraph object. On a miss, return None.
+        """
+        subdir = FxGraphCache._get_subdir(key)
+        if not os.path.exists(subdir):
+            return None
+
+        # Iterate over any entries in the subdir for this key and evaluate
+        # their guards to determine whether there's a hit.
+        for path in os.listdir(subdir):
+            with open(os.path.join(subdir, path), "rb") as f:
+                graph: CompiledFxGraph = pickle.load(f)
+
+            if not graph.guards_expr:
+                # No guards to evaluate
+                return graph
+
+            # Evaluate the guard expression in the current context.
+            tracing_context = torch._guards.TracingContext.get()
+            assert tracing_context is not None
+            shape_env = tracing_context.fake_mode.shape_env
+            symints = _filter_symints(example_inputs)
+
+            # We don't want to create the guards in the current shape env
+            # unless we have a hit.
+            with shape_env.suppress_guards():
+                res = shape_env.evaluate_guards_expression(graph.guards_expr, symints)
+                hit = definitely_true(res)
+
+            log.debug(
+                "fx graph evaluating guards for key %s: (%s) = %s",
+                key,
+                graph.guards_expr,
+                hit,
+            )
+            if hit:
+                # Now re-evaluate to add the guards to the current shape env.
+                # We have to clear the `evaluate_expr` lru_cache to force evaluation.
+                shape_env.evaluate_expr.cache_clear()
```

Comment:
Please comment. I don't know if clearing the whole cache is problematic, e.g., from a performance perspective. I could, for example, introduce a new context manager analogous to `suppress_guards()` above that selectively uses a cached vs. non-cached version of `shape_env.evaluate_expr`, but I thought it was worth asking if that's overkill.

Reply:
Seems off to me that we would clear cache. leaving this to @ezyang to comment

Reply:
I agree it seems kinda weird that we would need to clear the cache. If a cache hit is preventing you from adding a guard, then that should mean that that guard was already evaluated, no?

Reply:
See some of the comments above. The approach here is to:
1. Load a possible match from the cache.
2. Evaluate the guards, but in a mode that _does not_ modify the guards in the current environment, because: in the case that there's a miss, we don't actually want the current env to change.
3. If there's a hit, only then re-evaluate to cause the guards to be added.

The problem here is that the caching at `evaluate_expr()` interferes with #3

Reply:
@masnesral I think a better solution would be to customize the caching on `evaluate_expr` so that it respects whether or not guards were suppressed or not. WDYT?

Reply:
Hmm, well I guess it didn't matter when the hints were ints but it does matter if the hints are symints. I mean, I wouldn't be opposed to just turning off this caching entirely when you're playing funny tricks with symint hints.